InnerEye/ML/metrics.py

```diff
@@ -242,14 +242,23 @@ def calculate_metrics_per_class(segmentation: np.ndarray,
                          f"the label tensor indicates that there are {number_of_classes - 1} classes.")
     binaries = binaries_from_multi_label_array(segmentation, number_of_classes)
 
-    all_classes_are_binary = [is_binary_array(ground_truth[label_id]) for label_id in range(ground_truth.shape[0])]
-    if not np.all(all_classes_are_binary):
+    binary_classes = [is_binary_array(ground_truth[label_id]) for label_id in range(ground_truth.shape[0])]
+
+    # If ground truth image is nan, then will not be used for metrics computation.
+    nan_images = [np.isnan(np.sum(ground_truth[label_id])) for label_id in range(ground_truth.shape[0])]
```

Comment:
thinking through this piece of code here: We are going over the HUGE tensor multiple times, quite likely going out of the range of CPU cache. It would probably be more efficient if we did all checks in the loop in line 259ff, always only looking at one of the predicted structures at a time.